### ISSUE TITLE
sql: known invalid casts should consistently fail

### ISSUE THREAD

**Issue by maddyblue** [MaterializeInc/materialize#27172]
```
create table t (item int);
select * from t where item = 'word';
insert into t values (0);
select * from t where item = 'word';
```

The first `SELECT` succeeds with 0 rows, the second fails with a cast error. Postgres fails on the first query. The explain plan of the select with 0 rows in the table shows the cast error, so *something* is able to know about this beforehand. We should probably do something closer to postgres, or at the least actually do what our explain plan says we will do?

**Comment by maddyblue** [comment]
Can we add an opt step that pre-computes fns with constant args? Is that safe to do from an error semantics perspective (i.e., it may be wrong to have the optimizer in charge of whether certain queries produce errors)? Is there some other semantically safe place we could do that instead?

**Comment by maddyblue** [comment]
From @frankmcsherry 

> My sense is that it's maybe less an "opt step" and something closer to SQL -> HIR. We would only want to do it with casts, because e.g. select 1/0 from foo has a function of constant data, but we shouldn't evaluate eagerly.

**Comment by maddyblue** [comment]
Postgres indeed errors with CASE:

```
=# select * from t where case when i = 0 then i = 'word' else false end;
LOG:  00000: statement: select * from t where case when i = 0 then i = 'word' else false end;
LOCATION:  exec_simple_query, postgres.c:1013
ERROR:  22P02: invalid input syntax for type integer: "word"
LINE 1: select * from t where case when i = 0 then i = 'word' else f...
```

**Comment by benesch** [comment]
I don't think we need to do anything here! Evaluation order is not specified, and therefore we have tremendous latitude in determining whether a given query produces an error or not. 

> We should probably do something closer to postgres

I think trying to go error-for-error with PostgreSQL is a can of worms. It will be a forever game of whack a mole. We might be able to better align some of the simple cases, but in the limit, whether PostgreSQL produces an error on a given query depends heavily on the plan used, which in turn depends heavily on the exact version of PostgreSQL and the statistics of the tables involved in the query. 

I think the best approach here is simply to say that there Materialize no defined evaluation order (except for a carveout around `CASE`), and we don't promise to match PostgreSQL, nor use the same evaluation order from version to version.

> or at the least actually do what our explain plan says we will do?

The explain plan doesn't say the query will produce an error! It says it will apply a filter whose expression errored:

```
materialize=> explain select * from t where item = 'word';
                                           Optimized Plan                                           
----------------------------------------------------------------------------------------------------
 Explained Query:                                                                                  +
   Filter error("invalid input syntax for type integer: invalid digit found in string: \"word\"")  +
     ReadStorage materialize.public.t                                                              +
                                                                                                   +
 Source materialize.public.t                                                                       +
   filter=(error("invalid input syntax for type integer: invalid digit found in string: \"word\""))+
                                                                                                   +
 Target cluster: quickstart                                                                        +
 
(1 row)
```

And indeed, applying an errored filter on zero rows does not need to produce an error. 

> Can we add an opt step that pre-computes fns with constant args? 

Could do!

> Is that safe to do from an error semantics perspective (i.e., it may be wrong to have the optimizer in charge of whether certain queries produce errors)? Is there some other semantically safe place we could do that instead?

Yes, it's always safe. As long as there was some evaluation that _could_ have produced the error, it is always safe to change Materialize to produce that error.

----

I'm not sure that eagerly constant folding yields the desired behavior, though. It seems to break as many cases as it fixes. E.g., if you did eager constant folding, `SELECT 1/0 WHERE FALSE` would produce an error, whereas today in Materialize it does not, which is plausibly the more desirable behavior.

----

To summarize:

  * I would strongly discourage us from attempting to go error-for-error with PostgreSQL. We need the flexibility to reorder query evaluation however we see fit.
  * There may be a different good motivation for doing constant folding more eagerly. It doesn't seem like a slam dunk, though.

**Comment by maddyblue** [comment]
Agreed, thanks for the added context.
